### PR TITLE
Update access to use token and store firebaseInstanceId

### PR DESCRIPTION
**Pull request by xp-vit**

### CODE REVIEW

File: src/service/userService.js [hunk 1]
```diff
@@ -8,13 +8,77 @@ var backwardCompatibility = require("../util/backwardCompatibility");
...
+      user.firebaseInstanceIds = [];
+    }
+      async.detect(user.firebaseInstanceIds, (instanceIdTest, callback) => {
+      callback(null, instanceIdTest.instanceId === firebaseInstanceId);
+    }, (err, instanceIdFound) => {
+      if (err){
+        logger.log(err);
+      }
+      if(instanceIdFound){
+        logger.debug("Activating firebaseInstanceId: ", firebaseInstanceId, " for user: ", user.email);
```
Comment (dimhold) [inline src/service/userService.js]: Please use following format to be consistent:
```
logger.debug("[userService.addOrUpdateFirebaseInstanceId], "Activating firebaseInstanceId:", firebaseInstanceId, " for user:", user.email);
```

File: src/service/userService.js [hunk 2]
```diff
@@ -8,13 +8,77 @@ var backwardCompatibility = require("../util/backwardCompatibility");
 var passwordUtil = require("../util/password");
 
 module.exports = {
-  destroyAuthToken: function (user, callback) {
+  addOrUpdateFirebaseInstanceId (user, firebaseInstanceId, resultCallback) {
```
Comment (dimhold) [inline src/service/userService.js]: I suggest to rename
```
resultCallback -> callback
```
and async callback:
```
callback -> done
```
to be consistent

File: src/service/userService.js [hunk 3]
```diff
@@ -8,13 +8,77 @@ var backwardCompatibility = require("../util/backwardCompatibility");
...
+    }, (err, instanceIdFound) => {
+      if (err){
+        logger.log(err);
+      }
+      if(instanceIdFound){
+        logger.debug("Activating firebaseInstanceId: ", firebaseInstanceId, " for user: ", user.email);
+        instanceIdFound.lastUsedDate = Date.now();
+        instanceIdFound.active = true;
+      } else {
+        user.firebaseInstanceIds.push( { instanceId: firebaseInstanceId, active: true, createdDate: Date.now(), lastUsedDate: Date.now() } );
```
Comment (dimhold) [inline src/service/userService.js]: What happened if async trigger this callback with `err`?
I guess we execute else branch with:
`instanceId: undefined`

File: src/service/access.js [hunk 4]
```diff
@@ -2,17 +2,44 @@ var Errors = require("../error/errors");
...
 
 function checkAccess (req, res, next) {
-  var ip = req.ip;
-  var token = req.query.token;
-  logger.debug("[access.checkAccess] start. token: ", token, " ip: ", ip);
-  if (!token) {
-    logger.debug("[access.checkAccess] No Token. Send Unauthorized");
+  var ip = req.headers["x-forwarded-for"] || req.connection.remoteAddress;
+
+  var authHeader = req.headers.authorization;
```
Comment (dimhold) [inline src/service/access.js]: We use this method and `var firebaseInstanceId = req.get("FirebaseInstanceId");` in the same time. We should be consistent.

File: src/service/access.js [hunk 5]
```diff
@@ -2,17 +2,44 @@ var Errors = require("../error/errors");
...
+  logger.debug("[access.checkAccess] start. Authorization header: ", authHeader, " ip: ", ip);
+  if (!authHeader) {
+    logger.debug("[access.checkAccess] Authorization header. Send Unauthorized");
     sendUnauthorized(res);
     return;
   }
 
+  var authHeaderSplit = authHeader.split(" ");
```
Comment (dimhold) [inline src/service/access.js]: I guess we have very strict processing client request here. Client can send tabs, many spaces and as for me, it should be ok

File: src/service/access.js [hunk 6]
```diff
@@ -25,14 +52,21 @@ function checkAccess (req, res, next) {
       return;
     }
 
-    logger.debug("[access.checkAccess] Log in: ", user.email , " <== ", token);
+    var firebaseInstanceId = req.get("FirebaseInstanceId");
+    userService.addOrUpdateFirebaseInstanceId(user, firebaseInstanceId, function (err, user) {
+    if(!err){
+      req.firebaseInstanceId = firebaseInstanceId;
```
Comment (dimhold) [inline src/service/access.js]: Do we really need this line?

File: src/service/access.js [hunk 7]
```diff
@@ -25,14 +52,21 @@ function checkAccess (req, res, next) {
...
+      logger.debug("[access.checkAccess] Log in: ", user.email , " <== ", token," FirebaseInstanceId: ", firebaseInstanceId);
+    }
     updateIp(user, ip);
+    db.user.update(user);
     req.user = user;
+    req.ip = ip;
```
Comment (dimhold) [inline src/service/access.js]: Do we really need this line?

File: src/service/userService.js [hunk 8]
```diff
@@ -8,13 +8,73 @@ var backwardCompatibility = require("../util/backwardCompatibility");
...
+    return callback(null, user);
+  });
+},
+
+  destroyAuthToken (user, firebaseInstanceId, callback) {
     user.authToken = "";
-    db.user.update(user);
-    callback(null, {command: "logout", result: "done"});
+    this.deactivateFirebaseInstanceId(user, firebaseInstanceId, function (err, user) {
+      if (err && firebaseInstanceId) {
```
Comment (dimhold) [inline src/service/userService.js]: Can we just use simpler `err` check here?
```
if (err) {
...
```

File: src/service/userService.js [hunk 9]
```diff
@@ -202,28 +292,30 @@ verifyGoogleAndFindOrCreateUser: function (email, familyName, token, ip, callbac
...
-        var json = JSON.parse(googleJson);
+        json = JSON.parse(googleJson);
       } catch (e) {
         logger.warn("[userService.verifyGoogleAndFindOrCreateUser, ", email, "] Bad JSON: ", e.message);
-        callback(Errors.GoogleError());
-        return;
+        return callback(Errors.GoogleError());
       }
       logger.debug("[userService.verifyGoogleAndFindOrCreateUser, ", email, "] Recive json: ", json);
       if (json.family_name = familyName) {
```
Comment (dimhold) [inline src/service/userService.js]: :grimacing:

File: src/service/userService.js [hunk 10]
```diff
@@ -303,26 +415,33 @@ findOrCreateByGoogleData: function (id, email, ip, callback) {
...
+      });
     } else {
       logger.debug("[userService.findOrCreateByGoogleData, ", email, " User not exist. Try create him");
 
-      var user = {
+      var newUser = {
         authToken: crypto.randomBytes(config.app.tokenLength).toString("hex"), 
         googleId: id, 
         email: email,
         ip: ip
```
Comment (dimhold) [inline src/service/userService.js]: We don't have `firebaseInstanceIds : []` property on `newUser`. Is it ok?

File: src/util/password.js [hunk 11]
```diff
@@ -1,15 +1,16 @@
...
 var crypto = require("crypto");
+var config = require("config");
 
 module.exports = {
-  generateHashForPassword: function (email, password, sold) {
+  generateHashForPassword (email, password) {
     logger.data("[password.generateHashForPassword, ", email, "] Try generate hash.");
     var sha1sum = crypto.createHash("sha1");
-    sha1sum.update(password + email + sold);
+    sha1sum.update(password + email + config.app.secret);
```
Comment (dimhold) [inline src/util/password.js]: We start using global config varibale in utility. As a result we also updating config in tests... I think we should still have `sold` for `generateHashForPassword` and add sold arg for `isPasswordCorrect`

File: src/service/userService.js (hunk 10 above)
Comment (xp-vit) [inline src/service/userService.js]: Yes, it's ok. We have discussed offline.

File: src/service/access.js (hunk 4 above)
Comment (xp-vit) [inline src/service/access.js]: it's ok. We have discussed offline.

File: src/service/userService.js (hunk 9 above)
Comment (xp-vit) [inline src/service/userService.js]: Legacy issue )))

File: test/service/userServiceTest.js [hunk 12]
```diff
@@ -8,31 +8,31 @@ var passwordUtil = require("../../src/util/password");
 describe("User service.", function () {
   describe("Find or create user by login and password.", function () {
     it("Should return error when email is invalid", function () {
-      userService.findOrCreateByLoginAndPassword("this is not email", "", "127.0.0.1", function (err) {
+      userService.findOrCreateByLoginAndPassword("this is not email", "", "127.0.0.1", "FireBaseInstanceId", function (err) {
```
Comment (dimhold) [inline test/service/userServiceTest.js]: Will be fixed in next pull request

**Pull request merged**